cookbook/security/throttle_failed_login.rst

```diff
@@ -0,0 +1,117 @@
+.. index::
+    single: Security; Throttle Failed Login
+
+How to Throttle Failed Login Attempts
+=====================================
+
+Sometimes, it's useful to throttle login attempts when you encounter multiple
+recent failed logins.  This can be useful in combatting brute-force
+password-guessing login attacks.  This can be done by implementing a
+``security.authentication.failure`` listener to log failed login attempts, and
+a security provider decorator to deny login attempts from the same IP address.
+
+Authentication Failure Listener
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The purpose of this authentication failure listener is to persist the ip address
+and any other data of the client.  The example here uses a doctrine table to
+persist the data, but you could use another method (Redis, for example).
+
+.. code-block:: php
+
+    namespace AppBundle\Security\Authentication\Listener;
+
+    use Doctrine\ORM\EntityRepositoryInterface;
+    use Symfony\Component\HttpFoundation\RequestStack;
+    use Symfony\Component\Security\Core\Event\AuthenticationFailureEvent;
+
+    class LogFailedLoginAttempt
+    {
+        /**
+         * The doctrine repository to which the failed login attempt will be
+         * persisted.
+         *
+         * @var EntityRepositoryInterface
+         */
+        private $repository;
+
+        /**
+         * The current request contains the ip address and other $_SERVER globals
+         * to persist.
+         *
+         * @var RequestStack
+         */
+        private $requestStack;
+
+        public function __construct(RequestStack $requestStack, EntityRepositoryInterface $repository)
+        {
+            $this->repository = $repository;
+            $this->requestStack = $requestStack;
+        }
+
+        public function onAuthenticationFailureEvent(AuthenticationFailureEvent $event)
+        {
+            $this->repository->logAuthenicationFailure(
+                $event->getToken(),
+                $this->requestStack->getCurrentRequest()
+            );
+        }
+    }
+
+Now you are logging all the relevant details about every failed login attempt.
```

Comment:
I think it'll be easier to follow if you inject the EntityManager, then fetch some fake "AppBundle:LoginAttempt" entity (or something like that). Then we should also show this registered as a service. Finally, we can add a note that we'll leave the LoginAttempt entity logic code to the user.